mindmeld/augmentation.py

```diff
@@ -0,0 +1,285 @@
+from .constants import _get_pattern
+from .models.helpers import register_augmentor
+from .resource_loader import ResourceLoader
+
+logger = logging.getLogger(__name__)
+
+try:
+    import torch
+except ImportError:
+    logger.info(
+        "Library not found: 'torch'. Run 'pip install mindmeld[augment]'" " to install."
+    )
+
+try:
+    from transformers import PegasusForConditionalGeneration, PegasusTokenizer
+    from transformers import MarianMTModel, MarianTokenizer
+except ImportError:
+    logger.info(
+        "Library not found: 'transformers'. Run 'pip install mindmeld[augment]'"
+        " to install."
+    )
+
+# pylint: disable=R0201
+
+
+class Augmentor(ABC):
+    """
+    Abstract Augmentor class.
+    """
+
+    def __init__(self, app_path, config=None):
+        """Initializes an augmentor.
+
+        Args:
+            app_path (str): The location of the MindMeld app
+            config (dict, optional): A config object to use. This will
+                override the config specified by the app's config.py file.
+        """
+        self.app_path = app_path
+        self.config = config or get_augmentation_config(app_path=app_path)
+        self._resource_loader = ResourceLoader.create_resource_loader(app_path)
+
+    def augment(self, config, **kwargs):
+        """Augments queries given initial queries in application.
+
+        Args:
+            config (dict, optional): App config to use instead of class config.
+        """
+        config = config or self.config
+        filtered_paths = self._get_files(config)
+
+        for path in tqdm(filtered_paths):
+            queries = self._read_path_queries(path)
+            augmented_queries = self.augment_queries(queries, **kwargs)
+            self._write_files(path, augmented_queries)
+
+    @abstractmethod
+    def augment_queries(self, queries):
+        """Generates augmented data given application queries.
+
+        Args:
+            queries (list): List of queries.
+
+        Return:
+            augmented_queries (list): List of augmented queries.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+    def _get_files(self, config):
+        """Fetches relevant files given the path rules specified in the config.
+
+        Args:
+            config (dict): Application config.
+
+        Return:
+            filtered_paths (list): List of file paths to be augmeted.
+        """
+        all_file_paths = self._resource_loader.get_all_file_paths()
+
+        if not config["paths"]:
+            logger.warning(
+                """'paths' field is not configured or misconfigured in the `config.py`.
+                 Can't find files to augment."""
+            )
+            return
+
+        rules = config["paths"]
+
+        filtered_paths = []
+
+        for rule in rules:
+            pattern = _get_pattern(rule)
+            compiled_pattern = re.compile(pattern)
```

Comment:
I would first validate the patterns before renumerating the paths since we can pick up invalid patterns earlier

Reply:
The invalid patterns should be removed in the filtering step that follows.